### PR TITLE
Add action to sync release candidate fixes to master

### PR DESCRIPTION
**Pull request by dime10**
This workflow will sync changes in the release candidate branch to master during a feature freeze. The action is run daily and will only create a pull request when there exists a release candidate branch corresponding to the current package version as declared in `pennylane/_version.py`, that is if the branch "v0.23.0-rc0" exists while the package version is "v0.23.0-dev" (for example).

**Edit**: The action now assumes that a release candidate branch one version below the current dev version exists only during a feature freeze, that is the branch "v0.22.0-rc0" exists once the package version is bumped to "v0.23.0-dev" (start of the feature freeze), and is deleted at the end of the feature freeze ("v0.22.0" is released).

**Note**: Delete last rc branch before merging.

### CODE REVIEW

File: .github/workflows/rc_sync.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+      # Checks-out the PennyLane repository
+      - uses: actions/checkout@v3
+
+      # Check for the existence of an rc branch for the current PennyLane version
+      - name: Check for rc branch
+        run: |
+          VERSION=$(python setup.py --version)
+          if git ls-remote --quiet --exit-code origin "refs/heads/v${VERSION%.dev0}-rc0"; then
```
Comment (dime10) [inline .github/workflows/rc_sync.yml:34]:
```suggestion
          VERSION=$(python setup.py --version)
          IFS=. read MAJ MIN PAT <<< ${VERSION%.dev0}
          if git ls-remote --exit-code origin "refs/heads/v${MAJ}.$((MIN-1)).${PAT}-rc0"; then
```

File: .github/workflows/rc_sync.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+      # Create a new branch at from the current rc to be used in the PR
+      - name: Create temp branch
+        if: ${{ env.branch_exists == 'true' }}
+        run: |
+          git fetch
+          git checkout "${{ env.rc_branch }}"
+          BRANCH="rc_$(date +'%Y-%m-%d-%H-%M-%S')"
+          echo "tmp_branch=$BRANCH" >> $GITHUB_ENV
+          git checkout -b "$BRANCH"
```
Comment (dime10) [inline .github/workflows/rc_sync.yml:49]:
```suggestion
          git checkout -b "$BRANCH"
          git checkout master -- pennylane/_version.py
          git checkout master -- doc/releases/changelog-dev.md
          git commit -m "exclude files from pr"
```

File: .github/workflows/rc_sync.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      # Checks-out the PennyLane repository
+      - uses: actions/checkout@v3
+
+      # Check for the existence of an rc branch for the current PennyLane version
+      - name: Check for rc branch
+        run: |
+          VERSION=$(python setup.py --version)
+          if git ls-remote --quiet --exit-code origin "refs/heads/v${VERSION%.dev0}-rc0"; then
+            echo "branch_exists=true" >> $GITHUB_ENV
+            echo "rc_branch=v${VERSION%.dev0}-rc0" >> $GITHUB_ENV
```
Comment (dime10) [inline .github/workflows/rc_sync.yml:36]:
```suggestion
            echo "rc_branch=$RC_BRANCH" >> $GITHUB_ENV
```

File: .github/workflows/rc_sync.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        if: ${{ env.branch_exists == 'true' }}
+        run: |
+          git fetch
+          git checkout "${{ env.rc_branch }}"
+          BRANCH="rc_$(date +'%Y-%m-%d-%H-%M-%S')"
+          echo "tmp_branch=$BRANCH" >> $GITHUB_ENV
+          git checkout -b "$BRANCH"
+          git checkout master -- pennylane/_version.py
+          git checkout master -- doc/releases/changelog-dev.md
+          git commit -m "exclude files from pr"
```
Comment (dime10) [inline .github/workflows/rc_sync.yml:54]:
```suggestion
          if ! git diff-index --quiet HEAD; then
            git config user.name "GitHub Actions Bot"
            git config user.email "<>"
            git commit -m "exclude files from pr"
          fi
          git commit -m "exclude files from pr"
```

File: .github/workflows/rc_sync.yml [hunk 5]
```diff
@@ -0,0 +1,73 @@
+# This workflow will sync changes in the release candidate branch to master during a
+# feature freeze. The action is run daily and will only create a pull request when
+# there exists a release candidate branch corresponding to the current package version
+# as declared in pennylane/_version.py, that is if the branch "v0.23.0-rc0" exists while
+# the package version is "v0.23.0-dev".
```
Comment (antalszava) [inline .github/workflows/rc_sync.yml:5]:
```suggestion
# the package version is "v0.24.0-dev".
```

File: .github/workflows/rc_sync.yml [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+name: Sync rc with master
```
Comment (antalszava) [inline .github/workflows/rc_sync.yml:7]:
```suggestion
name: Sync release candidate with master
```

File: .github/workflows/rc_sync.yml [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+      # Checks-out the PennyLane repository
+      - uses: actions/checkout@v3
+
+      # Check for the existence of an rc branch for the current PennyLane version
+      - name: Check for rc branch
+        run: |
+          VERSION=$(python setup.py --version)
+          IFS=. read MAJ MIN PAT <<< "${VERSION%.dev0}"
```
Comment (antalszava) [inline .github/workflows/rc_sync.yml:34]: Does this line produce what we'd expect?

Executing it and hitting `echo $IFS`, there's no output.

Also, executing part of it, I'm not sure we get expected outputs here:
```shell
(base) antal@pc:~/xanadu/pennylane$ read MAJ MIN PAT <<< "${VERSION%.dev0}"
(base) antal@pc:~/xanadu/pennylane$ echo $MAJ
0.23.0
(base) antal@pc:~/xanadu/pennylane$ echo $MIN

(base) antal@pc:~/xanadu/pennylane$ echo $PAT
```
Maybe I'm doing something wrong though.

File: .github/workflows/rc_sync.yml [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      - name: Create temp branch
+        if: ${{ env.branch_exists == 'true' }}
+        run: |
+          git fetch
+          git checkout "${{ env.rc_branch }}"
+          BRANCH="rc_$(date +'%Y-%m-%d-%H-%M-%S')"
+          echo "tmp_branch=$BRANCH" >> $GITHUB_ENV
+          git checkout -b "$BRANCH"
+          git checkout master -- pennylane/_version.py
+          git checkout master -- doc/releases/changelog-dev.md
```
Comment (antalszava) [inline .github/workflows/rc_sync.yml:53]: Worth leaving a comment here on why we're doing this. :+1:

File: .github/workflows/rc_sync.yml [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+          if ! git diff-index --quiet HEAD; then
+            git config user.name "GitHub Actions Bot"
+            git config user.email "<>"
+            git commit -m "exclude files from pr"
+          fi
+          git commit -m "exclude files from pr"
```
Comment (antalszava) [inline .github/workflows/rc_sync.yml:59]: How come we're committing here?

**Review by antalszava — COMMENTED** [review]
The workflow is looking great, this will be very helpful! :star: 

Was wondering about some details, once addressed this would be good to go from my side. Would the simplest way to try this out be to merge and trigger the action?

File: .github/workflows/rc_sync.yml (hunk 9 above)
Comment (dime10) [reply .github/workflows/rc_sync.yml:59]: Oh sorry, the second commit should not be there, must have overlooked it when I suggested the change.

File: .github/workflows/rc_sync.yml (hunk 8 above)
Comment (dime10) [reply .github/workflows/rc_sync.yml:53]: Sounds good 👍

File: .github/workflows/rc_sync.yml [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+      # Create a new branch at from the current rc to be used in the PR
+      - name: Create temp branch
+        if: ${{ env.branch_exists == 'true' }}
+        run: |
+          git fetch
+          git checkout "${{ env.rc_branch }}"
+          BRANCH="rc_$(date +'%Y-%m-%d-%H-%M-%S')"
+          echo "tmp_branch=$BRANCH" >> $GITHUB_ENV
+          git checkout -b "$BRANCH"
+          git checkout master -- pennylane/_version.py
```
Comment (dime10) [inline .github/workflows/rc_sync.yml:52]:
```suggestion
          # Exclude some files known to be divergent during the release process from the PR
          git checkout master -- pennylane/_version.py
```

File: .github/workflows/rc_sync.yml (hunk 9 above)
Comment (dime10) [inline .github/workflows/rc_sync.yml:59]:
```suggestion
```

**Comment by dime10** [comment]
Thanks @antalszava for the review :)
I've fixed the minor issues you brought up. Regarding testing, I was testing the script in new repo I specifically created for that purpose, so it "should" work 😅 Once merged we can also test it on this repo by manually triggering it like you mentioned.

**Review by antalszava — APPROVED** [review]
Looks good! :tada:

**Comment by antalszava** [comment]
[sc-13147]

**Pull request merged**